src/github.com/matrix-org/dendrite/roomserver/types/types.go

```diff
@@ -163,10 +171,33 @@ type RoomRecentEventsUpdater interface {
 	HasEventBeenSent(eventNID EventNID) (bool, error)
 	// Mark the event as having been sent to the output logs.
 	MarkEventAsSent(eventNID EventNID) error
-	// Commit the transaction
-	Commit() error
-	// Rollback the transaction.
-	Rollback() error
+	// Build a membership updater for the target user in this room.
+	// It will share the same transaction as this updater.
+	MembershipUpdater(targetNID EventStateKeyNID) (MembershipUpdater, error)
+	// Implements Transaction so it can be committed or rolledback
+	Transaction
+}
+
+// A MembershipUpdater is used to update the membership of a user in a room.
+// (On postgresql this wraps a database transaction that holds a "FOR UPDATE"
+//  lock on the row in the membership table for this user in the room)
+type MembershipUpdater interface {
+	// True if the target user is invited to the room.
+	IsInvite() bool
+	// True if the target user is joined to the room.
+	IsJoin() bool
+	// True if the target user is not invited or joined to the room.
+	IsLeave() bool
+	// Set the state to invite.
+	// Returns whether this invite needs to be sent
+	SetToInvite(event gomatrixserverlib.Event) (needsSending bool, err error)
+	// Set the state to join.
+	SetToJoin(senderID string) (inviteIDs []string, err error)
+	// Set the state to leave.
+	// Returns a list of invite event IDs that this state change retired.
+	SetToLeave(senderID string) (inviteIDs []string, err error)
```

Comment:
Are these calls idempotent? Can I call `SetToLeave` twice and get the same IDs back without any errors? You need to either way in your interface.